File: POO.java
// Classe principal Veiculo
class Veiculo {
    // Atributos
    protected String marca;
    protected String modelo;
    protected int ano;
    protected boolean ligado;

    // Método construtor
    public Veiculo(String marca, String modelo, int ano) {
        this.marca = marca;
        this.modelo = modelo;
        this.ano = ano;
        this.ligado = false;
    }

    // Método para ligar o veículo
    public void ligar() {
        this.ligado = true;
        System.out.println(this.marca + " " + this.modelo + " ligado.");
    }

    // Método para desligar o veículo
    public void desligar() {
        this.ligado = false;
        System.out.println(this.marca + " " + this.modelo + " desligado.");
    }

    // Método para acelerar o veículo
    public void acelerar(int velocidade) {
        if (this.ligado) {
            System.out.println(this.marca + " " + this.modelo + " acelerando a " + velocidade + " km/h.");
        } else {
            System.out.println("Não é possível acelerar, " + this.marca + " " + this.modelo + " está desligado.");
        }
    }

    // Método para parar o veículo
    public void parar() {
        System.out.println(this.marca + " " + this.modelo + " parado.");
    }
}

// Classe Carro que herda de Veiculo
class Carro extends Veiculo {
    // Atributo específico de Carro
    private String tipo;

    // Método construtor
    public Carro(String marca, String modelo, int ano, String tipo) {
        super(marca, modelo, ano);
        this.tipo = tipo;
    }

    // Método para abrir o porta-malas
    public void abrirPortaMalas() {
        System.out.println(this.marca + " " + this.modelo + " abrindo o porta-malas.");
    }
}

// Classe Moto que herda de Veiculo
class Moto extends Veiculo {
    // Atributo específico de Moto
    private int cilindradas;

    // Método construtor
    public Moto(String marca, String modelo, int ano, int cilindradas) {
        super(marca, modelo, ano);
        this.cilindradas = cilindradas;
    }

    // Método para ligar o farol
    public void ligarFarol() {
        System.out.println(this.marca + " " + this.modelo + " ligando o farol.");
    }
}

// Classe Main
public class Main {
    public static void main(String[] args) {
        // Instâncias das classes Carro e Moto
        Carro carro = new Carro("Toyota", "Corolla", 2020, "Sedan");
        Moto moto = new Moto("Honda", "CG 125", 2019, 125);

        // Chamando métodos das classes Carro e Moto
        carro.ligar();
        carro.acelerar(60);
        carro.abrirPortaMalas();
        carro.parar();
        carro.desligar();

        moto.ligar();
        moto.acelerar(50);
        moto.ligarFarol();
        moto.parar();
        moto.desligar();
    }
}

File: POO.py
# Classe principal Veiculo
class Veiculo:
    # Método construtor
    def __init__(self, marca, modelo, ano):
        self.marca = marca
        self.modelo = modelo
        self.ano = ano
        self.ligado = False

    # Método para ligar o veículo
    def ligar(self):
        self.ligado = True
        print(f"{self.marca} {self.modelo} ligado.")

    # Método para desligar o veículo
    def desligar(self):
        self.ligado = False
        print(f"{self.marca} {self.modelo} desligado.")

    # Método para acelerar o veículo
    def acelerar(self, velocidade):
        if self.ligado:
            print(f"{self.marca} {self.modelo} acelerando a {velocidade} km/h.")
        else:
            print(f"Não é possível acelerar, {self.marca} {self.modelo} está desligado.")

    # Método para parar o veículo
    def parar(self):
        print(f"{self.marca} {self.modelo} parado.")


# Classe Carro que herda de Veiculo
class Carro(Veiculo):
    # Método construtor
    def __init__(self, marca, modelo, ano, tipo):
        super().__init__(marca, modelo, ano)
        self.tipo = tipo

    # Método para abrir o porta-malas
    def abrir_porta_malas(self):
        print(f"{self.marca} {self.modelo} abrindo o porta-malas.")


# Classe Moto que herda de Veiculo
class Moto(Veiculo):
    # Método construtor
    def __init__(self, marca, modelo, ano, cilindradas):
        super().__init__(marca, modelo, ano)
        self.cilindradas = cilindradas

    # Método para ligar o farol
    def ligar_farol(self):
        print(f"{self.marca} {self.modelo} ligando o farol.")


# Instâncias das classes Carro e Moto
carro = Carro("Toyota", "Corolla", 2020, "Sedan")
moto = Moto("Honda", "CG 125", 2019, 125)

# Chamando métodos das classes Carro e Moto
carro.ligar()
carro.acelerar(60)
carro.abrir_porta_malas()
carro.parar()
carro.desligar()

moto.ligar()
moto.acelerar(50)
moto.ligar_farol()
moto.parar()
moto.desligar()
